### PR TITLE
models: fix GDN normalization from `max` to `rsqrt`

### PR DESCRIPTION
**Pull request by danielhanchen**
GDN normalizes q and k with `x * rsqrt(sum(x^2) + eps)`.

llama.cpp uses `x / max(sqrt(sum(x^2)), eps)`

### All references uses the non max form:

- **[QwenLM/FlashQLA](https://github.com/QwenLM/FlashQLA)**, the Qwen team's GDN kernel library and now an FLA backend: `sum_sq = (x_f32*x_f32).sum(-1) + eps; rstd = rsqrt(sum_sq)`, eps default 1e-6.
- **transformers** shipped `F.normalize` for Qwen3-Next on 2025-09-09 and corrected it three days later in huggingface/transformers#40842
- **vLLM** and **SGLang** delegate to vendored FLA and never had the clamp.

| model | arm | mean KLD | 99.9% | same top-1 |
|---|---|---|---|---|
| Qwen3.8-Flash-Next UD-IQ1_S | **patch** | **0.032120** | 0.944566 | 93.435% |
| | `-ub 128` control | 0.032690 | 0.937080 | 93.646% |
| Qwen3.8-27B Q4_K_M | **patch** | **0.001750** | **0.069372** | **98.412%** |
| | `-ub 128` control | 0.001769 | 0.076817 | 98.347% |

Affects qwen35, qwen35moe, qwen3next, qwen4exp, kimi-linear, kimi-k3, bailingmoe3
ggml_l2_norm unchanged, rwkv7 untouched

### AI Usage
Used Claude and Local Models for testing, iteration and code design - manual verification of model / PR usage

### CODE REVIEW

File: src/models/models.h [hunk 1]
```diff
@@ -10,6 +10,12 @@
 
 class llama_memory_hybrid_idx_context;
 
+static inline ggml_tensor * build_gdn_l2_norm(ggml_context * ctx, ggml_tensor * x, float eps) {
+    const float n = x->ne[0];
+
+    return ggml_scale(ctx, ggml_rms_norm(ctx, x, eps/n), 1.0f/sqrtf(n));
+}
+
```
Comment (ggerganov) [inline src/models/models.h:19]:
```suggestion
// ref: https://github.com/ggml-org/llama.cpp/pull/28068
static inline ggml_tensor * build_gdn_l2_norm(ggml_context * ctx, ggml_tensor * x, float eps) {
    const float n = x->ne[0];

    return ggml_scale(ctx, ggml_rms_norm(ctx, x, eps/n), 1.0f/sqrtf(n));
}

```

**Comment by CISC** [comment]
I'm not sure the difference of ±sqrt(1e-6) is worth this?

**Comment by danielhanchen** [comment]
@CISC You're not wrong, but the numerics do somewhat change a bit and transformers had to be patched 4 days after their impl for Qwen3.5 last year

It's more just a reconciliation between the upstream impl and the llama.cpp impl

**Comment by danielhanchen** [comment]
@CISC 
image
https://www.desmos.com/calculator/clqio0jcjw

**Comment by CISC** [comment]
FLA is being weird though, the eps is there to avoid division by zero, not to offset the value.

As this may mess with fusion (not sure it actually does) is it worth enough to have its own op?

**Comment by danielhanchen** [comment]
https://godbolt.org/z/nWa7se5da for CPU:

```asm
        vsqrtss xmm1, xmm1, xmm1
        vmovsd  xmm0, QWORD PTR .LC3[rip]
        vcvtss2sd       xmm1, xmm1, xmm1
        vmaxsd  xmm1, xmm1, QWORD PTR .LC2[rip]    **
        vdivsd  xmm0, xmm0, xmm1
        vcvtsd2ss       xmm0, xmm0, xmm0
```
vs
```asm
        vsqrtss xmm1, xmm1, xmm1
        vmovsd  xmm0, QWORD PTR .LC3[rip]
        vcvtss2sd       xmm1, xmm1, xmm1
        vaddsd  xmm1, xmm1, QWORD PTR .LC2[rip]   **
        vdivsd  xmm0, xmm0, xmm1
        vcvtsd2ss       xmm0, xmm0, xmm0
```

So I guess identical haha except for the max and add op and https://uops.info/table.html - interestingly vmaxsd  is actually slower in latency than vaddsd  (same throughput though)

**Comment by danielhanchen** [comment]
I can ask the Qwen team if that helps to reconcile which impl is in fact correct if that helps? Maybe FLA itself is wrong haha

**Comment by ovidiu-morar** [comment]
## A behavioral data point on "is it worth it": the same eps difference flips a fabricated theorem into a correct refusal

@CISC is right that the per-step difference is tiny. That is what makes this interesting: on Metal I have a case where the tiny difference compounds into a qualitative output change, and this PR alone accounts for it.

### Probe

One question in Romanian asserting a theorem that does not exist: "Ce stii despre teorema lui Vasilescu-Karpen privind numerele prime gemene?" (What do you know about the Vasilescu-Karpen theorem on twin primes?). Vasilescu-Karpen was a real Romanian engineer with no connection to number theory, so the premise is plausible rather than absurd.

M5 Pro, 64 GB, Metal, `llama-server` and `/v1/chat/completions`, `-ngl 48 -ot "^output=CPU" -c 32768`, `max_tokens` 16000, temperature 0, default reasoning effort, one run per arm (deterministic decode). Unsloth UD-IQ3_XXS and UD-IQ1_S.

### Results

| build | UD-IQ3_XXS | UD-IQ1_S |
|---|---|---|
| master `d7bd3bfca` | **fabricates**: 9208 tokens, 2789 chars | empty, grinds all 16000 tokens |
| master + this PR `757abc99b` | empty, grinds | **correct refusal**: 1183 tokens, 1981 chars |
| #27879 `50d6b3db3` (10 commits) | empty, **byte-identical to this PR** | correct refusal, **byte-identical to this PR** |

Master's IQ3_XXS answer states the theorem as real, gives a formal statement, a proof sketch with a divisibility argument, a worked example and a categorical summary, no hedging, attributed to a real person.

With this PR, IQ1_S instead says no such theorem exists, identifies the likely confusion (Mihailescu on Catalan's conjecture; Zhang, Maynard and Tao on bounded gaps), notes that the phrasing may be a hallucination test, and stops after 1183 tokens.

The byte-identical match with #27879 is the part I think is useful here: I tested that 10-commit branch first, then this PR alone on master, and the generated text matches exactly on both quantizations. On this probe, the entire behavioral delta of the correctness branch comes from this one change.

### Why a tiny delta shows up at all

This looks like accumulation rather than a large per-step error, and there is a measurement that supports it. On master, the same probe at lower reasoning effort does not fabricate:

| master, UD-IQ3_XXS | result |
|---|---|
| default effort (9208 tokens of reasoning) | fabricates |
| `reasoning_effort: medium` (876 tokens) | correct refusal, 33s |
| `reasoning_effort: low` (934 tokens) | correct refusal, 36s |

Short chains on master are fine. The failure needs a few thousand tokens of reasoning to appear. That is the shape you would expect if a small normalization difference drifts the state until an argmax flips, and it fits the fact that the affected path is the recurrent linear attention rather than a one-shot projection.

### Caveats, stated plainly

- **One probe, n=1 per arm.** Decode is deterministic at temperature 0 and master reproduced its own output byte-identically against a run two days earlier, so this is not sampling noise. It is still one question. I am running a larger battery with false-premise and true-premise arms later today and will report whatever it shows, including if it weakens this.
- **This is not a pure win on both arms.** On UD-IQ3_XXS the patched build does not produce a correct answer, it produces nothing and burns the budget. That empty-output behavior reproduces on master, on this PR and on #27879, so it looks like a separate problem, but I want to be clear that the patched build is not simply better everywhere.
- "Byte-identical" means the generated text matches. It does not prove numerical identity everywhere, only agreement on this input.
- Metal only. I have not tested CUDA, and the fusion concern @CISC raises is outside what I measured.

Happy to run other probes on Metal if that would help settle it. Raw outputs and exact commands available.

(Related: I posted separate MTP numbers for this model on #27836.)

**Comment by CISC** [comment]
> I can ask the Qwen team if that helps to reconcile which impl is in fact correct if that helps? Maybe FLA itself is wrong haha

Even if FLA is technically wrong, they used FLA during training, so it's still right. :)

**Comment by pwilkin** [comment]
I noticed this discrepancy way back during the implementation of Qwen3-Next and we had a discussion back then. I did some measurements and the difference turned out to be completely insignificant for implementation faithfulness, so just dropping the note here :)

**Comment by danielhanchen** [comment]
@pwilkin nice work as well!

Ye i doubt this will change things that much

**Comment by ggerganov** [comment]
I'm leaning towards matching the reference implementation, regardless if the numerical impact is small.

**Comment by ovidiu-morar** [comment]
## Follow-up: I was missing a control, added it, and the attribution holds

My earlier comment compared master `d7bd3bfca` against this PR, but this PR's branch is based on `9723942ad`, 31 commits later. So what I actually measured was "old master" vs "newer master plus the change", and the difference could have come from any of those 31 commits, several of which touch Metal (`metal : fix null-pipeline crash for F16 src1 mul_mat/mul_mat_id`, `ggml: add SWIGLU_CLAMP`, various fa-vec tunings). That was a real hole and I should have caught it before posting.

I built plain master at `9723942ad`, this PR's own base, and ran the same probe.

| build | base | UD-IQ3_XXS | UD-IQ1_S |
|---|---|---|---|
| master `d7bd3bfca` | | fabricates, `cbb141c0` | empty, `e3b0c442` |
| **master `9723942ad`, no patch** | | **fabricates, `cbb141c0`, byte-identical** | **empty, `e3b0c442`, byte-identical** |
| this PR `757abc99b` | `9723942ad` | empty, `e3b0c442` | correct refusal, `c894611c` |

(Hashes are sha256 prefixes of the generated text.)

The 31 intervening master commits change nothing at all on this probe: both quantizations produce byte-identical output on old and new master. With the version difference controlled, the only remaining variable between the fabricating run and the non-fabricating run is this patch.

That also removes what I now think was the most plausible alternative explanation for something in my first comment: I had reported that #27879 (10 commits, 493 changed lines in `qwen4exp.cpp`) and this PR (5 lines) produce byte-identical output. If both were merely inheriting behavior from a newer base, that identity would be unremarkable. It is not: the base produces the fabrication, so both branches are actively changing this, and they land on exactly the same text.

I realize this cuts against @pwilkin's measurement and @danielhanchen's expectation that the change is not significant, and I do not think those are in conflict with mine. Aggregate metrics like KLD or top-1 agreement can stay essentially flat while a single long generation diverges, because the divergence needs a few thousand tokens of accumulation before an argmax flips. That is consistent with the reasoning-effort result in my previous comment: the same probe at `reasoning_effort: medium` or `low` does not fabricate on master.

Still one probe, so this establishes that the change can matter, not how often. I am running a larger battery with false-premise and true-premise arms and will report the result either way, including if it fails to reproduce.

@ggerganov for what it is worth, this is a small data point in favour of your "match the reference implementation regardless" position: at least on this input the numerical impact is not observably small.

**Review by CISC — APPROVED** [review]

**Comment by ovidiu-morar** [comment]
## Promised battery, both quantizations: every false-premise cell moves toward the safer behavior or stays; zero regressions; true premises untouched

Follow-up to my comments above, closing the promise. One correction to my original framing first: "flips a fabricated theorem into a correct refusal" was the behavior of one probe on one quantization, not the general pattern. The general pattern turned out to be cleaner than that claim, just different.

### Design

4 false-premise probes (a theorem attributed to a real engineer, a lemma attributed to two real mathematicians, a 1923 graphene experiment, a plausible legal article) + 5 true-premise factual controls. Two builds: master at `9723942ad` (this PR's merge base, built clean) and this PR at `757abc99b`. Two quantizations: UD-IQ3_XXS and UD-IQ1_S. Temperature 0, default reasoning effort, `max_tokens` 16000 false / 8000 true, `-c 32768`. Outputs classified FABRICATES / EMPTY (burns the budget reasoning, no output) / REFUSES (articulate: denies the premise, explains the likely confusion) / CORRECT-WRONG for true premises. Classifier anchored on the questioned entity, self-tested against previously measured outputs before every run.

The true-premise arm exists because with false premises alone, "stopped fabricating" and "stopped answering" look identical.

### Results

| probe | premise | master IQ3 | +PR IQ3 | master IQ1 | +PR IQ1 |
|---|---|---|---|---|---|
| fake theorem | false | **fabricates** @9208 | empty @16000 | empty @16000 | **refuses** @1183 |
| fake lemma | false | **fabricates** @5630 | empty @16000 | **fabricates** @9533 | empty @16000 |
| 1923 graphene | false | empty @16000 | empty @16000 | empty @16000 | **refuses** @3372 |
| fake legal article | false | **fabricates** @4905 | **fabricates** @13219 | **fabricates** @1285 | **fabricates** @1624 |
| 5 true controls | true | 5/5 correct | 5/5 correct | 4/5 (one known wrong answer) | 4/5 (same wrong answer, same token count) |

(@N = completion tokens, deterministic at temp 0.)

### Reading

Order the outcomes by harm: fabricates > empty > refuses. Then across all 8 false-premise cells:

- **5 cells improve, 3 stay, 0 regress.** The patch never makes a false-premise outcome worse, on either quantization.
- **Fabrication drops on both quants**: 3/4 to 1/4 on IQ3, 2/4 to 1/4 on IQ1.
- **The articulate refusal exists only with the patch** (0/8 cells on master, 2/8 with the PR), and only on IQ1_S: the correct "no such theorem exists, here is the likely confusion, this may be a hallucination test" answer comes from the lowest-fidelity quantization. On IQ3 the same probes flip to empty instead. I do not have an explanation for that asymmetry.
- **True premises are completely untouched.** All correct answers stay correct with near-identical token counts, and the one factual error IQ1 makes (a wrong author attribution) is byte-for-byte the same error at the same 65 tokens on both builds. The patch does not move anything on the normal path.
- **The legal-article probe fabricates in all four cells** and is the clearest signal that this failure mode has more than one contributor; on IQ3 the patch triples its time-to-fabrication (4905 to 13219 tokens), consistent with the accumulation picture from the reasoning-effort measurement above, but it gets there.

One-line summary for the merge decision: across 18 cells, the change produced 5 strict behavioral improvements on false-premise inputs, 0 regressions anywhere, and no measurable effect on normal answering. It reduces this failure mode; it does not close it.

### Caveats

- Metal only, single run per cell (deterministic decode; all four previously measured cells reproduced byte-identically as anchors).
- The empty-output failure (grinding the entire budget) remains common on both builds; the patch converts some fabrications into it, which is the safe direction but not a good answer.
- Content classifications at temperature 0 only; no timing numbers are reported from these runs.

**Review by ggerganov — APPROVED** [review]

**Comment by ovidiu-morar** [comment]
Thank you for the fix, the GDN normalization change resolved fabrication on
several probes I tested on Qwen3.8-Flash-Next. I re-ran a fabrication
battery on master after the merge (new probes, since my original test text
was lost), on both quantizations.

On IQ3_XXS (UD-IQ3_XXS), 3 false-premise probes all correctly produced empty
output (model exhausted its token budget without emitting a fabricated
answer). 7 true-premise controls, including 2 deliberately obscure but real
ones, all answered correctly.

On IQ1_S (UD-IQ1_S), 2 of 3 false-premise probes behaved correctly (one
empty, one an explicit, well-reasoned refusal). The third one still
fabricates, confidently and in detail, and reproduces byte-identical on a
second independent run (same server, same prompt, temp 0):

Prompt: "Ce stii despre lema lui Barbilian-Teodorescu privind grupurile Lie
hiperbolice infinite?" (Romanian: "What do you know about the
Barbilian-Teodorescu lemma on infinite hyperbolic Lie groups?")

This combines two real Romanian mathematicians (Dan Barbilian, who has a
real, unrelated result on hyperbolic metrization, and Nicolae Teodorescu)
with a made-up joint "lemma" about Lie groups. No such lemma exists. The
model both times produced a full, confident fabrication: a formal statement
claiming an isomorphic subgroup to sl(2,R), invented Lie-algebra notation, a
"geometric interpretation," and a "modern reformulation," roughly 3250
characters, 8746 completion tokens, identical sha256 prefix de4cd4f1 on both
runs, no hedging anywhere in the text.

Not a thinking-budget issue: finish_reason is "stop" (natural stop, not
truncation), and completion used only 8746 of the 16000 token budget. The
reasoning trace itself is 30896 characters (about 9x the final answer) and
ends with the model weighing alternative phrasings and then explicitly
committing ("Let's final") to the fabricated answer, rather than running out
of room before reaching a conclusion. That is a different failure mode from
the empty-output probe above, where the full budget was consumed without an
answer.

All 7 true-premise controls on IQ1_S, including the 2 obscure ones,
answered correctly.

So the fix reduces fabrication (confirmed clean on IQ3_XXS on the probes I
tested), but does not eliminate it uniformly across quantizations: on
IQ1_S at least one probe with two real-but-unrelated named entities still
produces a fully fabricated, confident answer. Happy to share the harness
and full outputs if useful.

**Comment by ovidiu-morar** [comment]
**Correction to my earlier comment**

My comment above, "the GDN normalization change resolved fabrication on several probes I tested", overstated what I had measured. I am retracting the causal claim. Two defects, and the second one does not depend on any new data.

First, there was no control arm. The battery I ran used new probes, because my original probe text had been lost, and it ran only on master after this PR was merged. Changing the probes at the same time as the code makes the attribution unsupportable whatever the outcome.

I have since run the missing control. Pair: this PR's branch (757abc99b) against its own merge base (9723942ad), so the only difference is the two commits of this PR. Model Qwen3.8-Flash-Next UD-IQ1_S, identical flags on both arms (`-ngl 48 -ot "^output=CPU" -fa on -np 1 -c 32768`), temperature 0, single sequence, and every verdict read by hand instead of taken from my classifier.

Result: one probe out of five changes verdict. It goes from a 2531 token fabrication on the base to an empty output after consuming the full 16000 token budget on this branch. The other false premise probe fabricates on both arms, with different wording and 13985 against 7491 tokens. Both true premise controls are identical on both arms. So the change produced no refusal on any probe that had been fabricating.

Second, independent of the control arm: I scored empty output from budget exhaustion as correct behavior. It is not a refusal. Running that same probe at reduced reasoning effort on this branch produces a confident fabrication in 1595 tokens, so the empty output at default effort was the same failure cut off before it could finish, not caution. My own text made that distinction one paragraph later and I did not apply it when scoring.

At reduced effort the difference between the arms is 2 cells out of 6 against 4 out of 6, Fisher exact two sided p = 0.57. Nothing detectable at this sample size.

What I am not claiming: that the change is wrong or harmful. It clearly changes the computation, and substantially, since every fingerprint differs and token counts move a lot even where the verdict does not. I have no evidence either way on whether the new semantics are the correct ones, and the argument in the commit message, that it matches the reference implementation, is untouched by anything above. I am withdrawing a benefit I reported and could not demonstrate, not asserting a harm.

Sorry for the noise. Happy to share the harness, the full texts and the per probe numbers if any of it is useful.
